File: cpp/testcode.cc
#include "NNPDFDriver.h"
#include "LHAPDF/LHAPDF.h"
#include <cmath>
#include <stdlib.h>
#include <iomanip>
using namespace std;

int main(int argc, char** argv)
{
  string gridname = "";
  int member = 0;
  if (argc > 2)
    {
      gridname = argv[1];
      member = atoi(argv[2]);
    }
  else
    {
      cout << "\nusage: ./testcode <gridname> <member>\n" << endl;
      exit(-1);
    }

  string xpdf[] = {"x*tbar","x*bbar","x*cbar","x*sbar","x*ubar","x*dbar",
		   "x*g","x*d","x*u","x*s","x*c","x*b","x*t","x*photon"};
  double xlha[] = {1e-7, 1e-6, 1e-5, 1e-4, 1e-3, 1e-2,
		   1e-1, 3e-1, 5e-1, 7e-1, 9e-1};  
  double Q2[] = { 1.0, 2.0, 10.0, 100.0, 1000.0, 10000.0 };
  
  // fast option load single grid file
  NNPDFDriver *nnpdf = new NNPDFDriver(gridname, member);

  // or slow option
  // NNPDFDriver *nnpdf = new NNPDFDriver(gridname);
  // nnpdf->initPDF(member);
  
#if LHAPDF_MAJOR_VERSION > 5
  const LHAPDF::PDF*pdf = LHAPDF::mkPDF(gridname, member);
#define XFS(X,Q,F) pdf->xfxQ(F,X,Q)
#define XFSPHT(X,Q,F) pdf->xfxQ(F,X,Q)
#else
  LHAPDF::initPDFSet(gridname);
  LHAPDF::initPDF(member);
#define XFS(X,Q,F) LHAPDF::xfx(X,Q,F)
#define XFSPHT(X,Q,F) LHAPDF::xfxphoton(X,Q,F)
#endif

  double sum = 0;
  int ntot = 6;
  if (nnpdf->hasPhoton()) ntot++;

  for (int f = -6; f <= ntot; f++)
    {
      for (int iq = 0; iq < 6; iq++)
	{	
	  cout << "  " << xpdf[f+6] << ", Q2 = " << Q2[iq] << endl;
	  cout << "   x \t     C++ \tLHAPDF  \tDiff" << endl;
	  for (int ix = 0; ix < 11; ix++)
	    {
	      double a = nnpdf->xfx(xlha[ix],sqrt(Q2[iq]), f);
	      double b = 0;
	      if (nnpdf->hasPhoton())
		b = XFSPHT(xlha[ix],sqrt(Q2[iq]),f);
	      else
		b = XFS(xlha[ix],sqrt(Q2[iq]),f);
	      cout << scientific << setprecision(1) << xlha[ix] << "  ";
	      cout << scientific << setprecision(5) << a << "  " << b << "  " << a-b << endl;
	      sum += a-b;
	    } 
	  cout << endl;
	}
    }

  cout << "Sum of differences... " << sum << endl;    
  
  delete nnpdf;

  return 0;
}

File: fortran/testcode.f

      program testcode
*
      implicit none
*
      integer member,ntot,f,iq,ix
      character*100 gridname
      character*100 xpdf(-6:7)
      double precision sum
      double precision Q2(5)
      double precision xlha(11)
      double precision pdf(-6:7),pdf2(-6:7) 
      logical has_photon
      external has_photon
      
      data xpdf / "x*tbar","x*bbar","x*cbar","x*sbar","x*ubar","x*dbar",
     1	   "x*g","x*d","x*u","x*s","x*c","x*b","x*t","x*photon" /
      data Q2 / 2d0, 1d1, 1d2, 1d3, 1d4 /
      data xlha / 1d-7, 1d-6, 1d-5, 1d-4, 1d-3, 1d-2,
     1     1d-1, 3d-1, 5d-1, 7d-1, 9d-1 /
      
      gridname = "NNPDF23_nlo_as_0119.LHgrid"
      member = 0

      call NNPDFDriver(gridname)      
      call NNinitPDF(member)
            
      call InitPDFSetByName(gridname)
      call initPDF(member)

      sum = 0d0
      ntot = 6
      if (has_photon().eqv..true.) ntot = ntot+1

      do f=-6,ntot
         do iq=1,5
            write(6,*) "Q2 = ", Q2(iq)
            write(6,*) xpdf(f)
            write(6,*) "   x      Fortran    LHAPDF"
            
            do ix=1,11
               if (has_photon().eqv..false.) then                  
                  call evolvePDF(xlha(ix),Q2(iq)**0.5,pdf)                  
               else
                  call evolvePDFphoton(xlha(ix),Q2(iq)**0.5,pdf,pdf(7))
               endif
               call NNevolvePDF(xlha(ix),Q2(iq)**0.5,pdf2)
               write(6,*) xlha(ix),pdf2(f),pdf(f)
               
               sum = pdf(f)-pdf2(f)
            enddo
            write(6,*) ""
         enddo
      enddo
      
      if (sum.ne.0d0) then
         write(6,*) "Problems in the interpolation...",sum
      endif

      stop
      end
